consensus/istanbul/uptime/monitor_test.go

```diff
@@ -83,22 +80,130 @@ func TestUptimeSingle(t *testing.T) {
+	assert.ErrorIs(t, err3, ErrWrongEpoch)
+}
+
+func TestMonitorAddFirstHeaderOfEpoch(t *testing.T) {
+	assert.True(t, istanbul.IsFirstBlockOfEpoch(101, 100))
+	monitor := NewMonitor(100, 2, 10, 3)
+	// First block from epoch
+	header101 := mockHeader(101, big.NewInt(7), common.Hash{})
+	err := monitor.ProcessHeader(header101)
+	assert.NoError(t, err)
+	assert.Equal(t, header101, monitor.GetLastProcessedHeader())
+}
+
+func TestMonitorNotComputeUntilLockbackWindowPlusOne(t *testing.T) {
+	var uptime []*big.Int
+	monitor := NewMonitor(100, 2, 2, 3)
+	// First block from epoch
+	header101 := mockHeader(101, big.NewInt(0), common.Hash{})
+	err := monitor.ProcessHeader(header101)
+	assert.NoError(t, err)
+	_, err = monitor.ComputeUptime(header101)
+	assert.ErrorIs(t, err, ErrUnpreparedCompute)
+	header102 := mockHeader(102, big.NewInt(0), header101.Hash())
+	err = monitor.ProcessHeader(header102)
+	assert.NoError(t, err)
+	_, err = monitor.ComputeUptime(header102)
+	assert.ErrorIs(t, err, ErrUnpreparedCompute)
+	header103 := mockHeader(103, big.NewInt(7), header102.Hash())
+	err = monitor.ProcessHeader(header103)
+	assert.NoError(t, err)
+	uptime, err = monitor.ComputeUptime(header103)
+	assert.NoError(t, err)
+	assert.Equal(t, []*big.Int{params.Fixidity1, params.Fixidity1, params.Fixidity1}, uptime)
+	assert.Equal(t, monitor.GetLastProcessedHeader(), header103)
+}
+
+func TestMonitorAddLastHeaderOfEpoch(t *testing.T) {
+	monitor := NewMonitor(5, 2, 2, 3)
+	// First block from epoch
+	header6 := mockHeader(6, big.NewInt(0), common.Hash{})
+	monitor.ProcessHeader(header6)
+	header7 := mockHeader(7, big.NewInt(7), header6.Hash())
+	monitor.ProcessHeader(header7)
+	header8 := mockHeader(8, big.NewInt(0), header7.Hash())
+	monitor.ProcessHeader(header8)
+	header9 := mockHeader(9, big.NewInt(0), header8.Hash())
+	monitor.ProcessHeader(header9)
+	header10 := mockHeader(10, big.NewInt(0), header9.Hash())
+	monitor.ProcessHeader(header10)
+	uptime, err := monitor.ComputeUptime(header10)
+	assert.NoError(t, err)
+	halfScore := big.NewInt(0).Div(params.Fixidity1, big.NewInt(2))
+	assert.Equal(t, []*big.Int{halfScore, halfScore, halfScore}, uptime)
+	assert.Equal(t, header10, monitor.GetLastProcessedHeader())
+}
+
+func TestMonitorAddSameHeader(t *testing.T) {
+	monitor := NewMonitor(100, 2, 10, 2)
+	// First block from epoch
+	header101 := mockHeader(101, big.NewInt(7), common.Hash{})
+	err := monitor.ProcessHeader(header101)
+	assert.NoError(t, err)
+	err = monitor.ProcessHeader(header101)
+	assert.ErrorIs(t, err, ErrHeaderNumberAlreadyUsed)
+}
+
+func TestMonitorAddGappedHeader(t *testing.T) {
+	monitor := NewMonitor(100, 2, 10, 2)
+	// First block from epoch
+	header101 := mockHeader(101, big.NewInt(7), common.Hash{})
+	err := monitor.ProcessHeader(header101)
+	assert.NoError(t, err)
+	header103 := mockHeader(103, big.NewInt(7), common.Hash{})
+	err = monitor.ProcessHeader(header103)
+	assert.ErrorIs(t, err, ErrMissingPreviousHeaders)
```

Comment:
This type of predefined value for an error is commonly referred to as a sentinel error.

I would suggest avoiding the use of them in this context for 3 reasons:

* They prevent addition of useful information that you would really want to know when an error occurs in production. In the case of this error, did it occur because the monitor was not initialised, or was it because the provided header is out of sequence, and in that case what is the header number and what is the previous number? None of that context can be added because the test is checking for this error explicitly.
* We don't actually need them for tests, because it's not important what error is actually returned in this situation. What's important is that if there is a gap in the headers then an error is returned. The tests verify that this is the case. If we wan't to understand what code paths were executed we can use coverage to know for sure.
* They make the tests brittle because the tests will now fail if a different error is returned, even though the behaviour that we are interested in (err when gap in header) has been maintained.

There is a great article [here](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) by dave cheney that explains in more detail the problems they can create.

Reply:
I agree and disagree on this one.
I agree that making sentinel errors shouldn't be something to adopt. But this is a unit test of a specific class. Making other packages to know about which error is what, is wrong, but making the package test, to know the errors, is part of understanding which case is which.
I want to know that I'm sending this particular error because I'm forcing the scenario where that is happening.
The article that Dave wrote (which I agree) is about handling error of different packages. I see the test as part of the package. The test should know what's exactly going on, because is testing those specific scenarios. Even if the consumers uses or not a Sentinel way, the tests have to understand precisely what happened.
And if something in the package changed the behavior of this particular test, is the obligation of the test, let you know that the particular case you where testing, is not the same. That dev, will have to read the test, and decide if should change the error, or if for some reason, it broke use case that didn't realize that it broke.

Reply:
Hi @gastonponti

> I want to know that I'm sending this particular error because I'm forcing the scenario where that is happening.

Why do we need to know? What does it change?

As an example we test with missing headers.
* Option A an error is returned
* Option B ErrMissingPreviousHeaders is returned

In both cases I know that this function will reject input with missing headers, that's all I need to know, how it rejected it is up to the implementation and that can change, but as long as the function rejects missing headers then we know it is following the contract that other packages expect of it. Why then introduce a more restrictive contract (to return specific functions) inside the package, what is gained?